internal/internal_workflow_client.go

```diff
@@ -1162,8 +1162,15 @@ func (wc *WorkflowClient) loadCapabilities() (*workflowservice.GetSystemInfoResp
 	}
 
 	// Fetch the capabilities
-	ctx, cancel := context.WithTimeout(context.Background(), getSystemInfoTimeout)
-	defer cancel()
+
+	// We set a default timeout if none is set in the context to stay
+	// backward compatible with the old behavior where the timeout was
+	// hardcoded to 5s.
```

Comment:
Hrmm, I am a bit concerned to remove 5s timeout just because context has a deadline. This call should really complete quick. Just because I have an outer context governing lots of things I may be doing and have set a timeout for 10m lets say, doesn't mean that I should have to wait 10m for this specific call to fail. I think we should add the timeout on top of the given context no matter what (i.e. a `min` of user deadline or 5s, still giving user cancelability).

IIRC the use case bringing about the PR was for human interaction in a gRPC interceptor, but I am not sure it's a fair default that we should allow this call to start taking as long as the outer context's deadline. Is there any way you can solve your use case without removing this timeout for everyone that just happens to have a deadline on their context?

Reply:
Yeah this is not a backwards compatible change

Reply:
Yeah this is not a backwards compatible change in its current form

Reply:
> IIRC the use case bringing about the PR was for human interaction in a gRPC interceptor, but I am not sure it's a fair default that we should allow this call to start taking as long as the outer context's deadline. Is there any way you can solve your use case without removing this timeout for everyone that just happens to have a deadline on their context?

Not anymore. Human interaction in interceptor is worked around by performing the interaction outside of the scope of the gRPC request. What I am trying to address here is slow/spotty networks. 

Some people operate very far from Temporal servers (basically the other side of the globe), some are working with ISPs with slow DNS resolution times, working from bad wifi connections from airports, hotels, ... working over slow VPNs, working in transit from trains over a spotty network, relying on slow 3G network, ... usually those are people on call trying to respond to incidents and willing to run some automations in Temporal for instance. Hitting a 5s timeout when you combine a couple of those situations and a bit of network retries is possible.

Couple of years ago, there was wifi network congestion in some offices for weeks also impacting latencies and people used to extend their `tctl` `--context_timeout` flag, I believe to control the overall RPC timeout before [loading server capabilities became a thing](https://github.com/temporalio/sdk-go/pull/706). That kind of flexibility was nice but we kinda lost it.

Sadly, we don't instrument developers laptops, I don't have metrics giving us the P99 dial latency for all our clients, nor traces to pinpoint where time is spent. This is based on couple of user reports that despite retries still hit the timeout.

5s timeout work 99% of the time but maybe 30s is a more appropriate default timeout in our situation. As you don't control environments clients run into having something customizable makes sense to me. Otherwise, your only option when facing timeouts are to work around the problem by ssh'ing into a node closer to server, fidgeting with auth, ... or try to find a better network.

If we take that stance, wether we can the overall dial timeout or every single RPC timeout triggered by the client dial call is a tradeoff. Expose simple imperfect configuration option or complex low level ones that compounds each time you add a new RPC call in the middle of the dial call.

Reply:
Sounds like you may need the ability to control the get-system-info timeout more directly instead of changing everyone's default of 5s to their outer context. Many use large contexts with large deadlines but definitely don't have an initial connection/bootstrapping timeout expectation of that.

In general if any call like this takes longer than 5s you're going to have a subpar Temporal experience. We have task timeouts and such on pollers and Temporal in general is not built with the expectation that very simply calls to the server can take many seconds.

Reply:
Yeah it seems like what we need is the option for users to explicitly set the timeout of `get-system-info`.

Reply:
> Maybe we can have a client option for this.

Sure thing, I can revert back to that!

> In general if any call like this takes longer than 5s you're going to have a subpar Temporal experience. We have task timeouts and such on pollers and Temporal in general is not built with the expectation that very simple calls to the server can take many seconds. You're going to hit task timeouts and delays everywhere with an over-5s latency.

For sure! The issue is purely for CLI tools triggering workflows asynchronously from engineers laptops, not for workers processing tasks. Workers are in the cloud doing their job in a timely manner.